appsec/src/extension/request_lifecycle.c

```diff
@@ -426,14 +433,146 @@ static zend_array *nullable _response_commit(
 
     mlog(dd_log_debug, "Committing user request for span %p", span);
 
+    zend_string *rbe = _get_entity_as_string(rbe_zv);
+
     zend_array *res = _do_request_finish_user_req(
-        false, _superglob_equiv, status, resp_headers);
+        false, _superglob_equiv, status, resp_headers, rbe);
+
+    if (rbe) {
+        zend_string_release(rbe);
+    }
 
     _shutdown_done_on_commit = true;
 
     return res;
 }
 
+static zend_string *nullable _get_entity_as_string(zval *rbe_zv)
+{
+    if (!rbe_zv) {
+        return NULL;
+    }
+
+    const size_t max_size = (size_t)get_DD_APPSEC_MAX_BODY_BUFF_SIZE();
+
+    if (Z_TYPE_P(rbe_zv) == IS_STRING) {
+        if (Z_STRLEN_P(rbe_zv) <= max_size) {
+            zend_string *res = Z_STR_P(rbe_zv);
+            zend_string_addref(res);
+            return res;
+        }
+        mlog(dd_log_debug,
+            "Response body entity is larger than %zu bytes (got %zu); ignoring",
+            max_size, Z_STRLEN_P(rbe_zv));
+        return NULL;
+    }
+
+    if (Z_TYPE_P(rbe_zv) != IS_RESOURCE) {
+        mlog(dd_log_debug,
+            "Response body entity is not a string or a stream; ignoring");
+        return NULL;
+    }
+
+    php_stream *stream = NULL;
+    php_stream_from_zval_no_verify(stream, rbe_zv);
+    if (!stream) {
+        mlog(dd_log_debug,
+            "Response body entity is not a string or a stream; ignoring");
+        return NULL;
+    }
+
+    // TODO: support non-seekable streams. Needs replacing the stream
+    if (stream->flags & PHP_STREAM_FLAG_NO_SEEK) {
+        mlog(dd_log_info, "Response body entity is a stream, but it is "
```

Comment:
Any idea when we'll have non-seekable streams? If common, setting this to info might not be a great idea.

Reply:
at the end the stream code was not needed for roadrunner because we end up instrumenting the roadrunner specific class `\Spiral\RoadRunner\Http\HttpWorker`. That only only takes a string and a generator of strings (the latter in `respondStream`, for support of chunked responses).

PSR-7 does define an interface with streams, but its implementation in RoadRunner delegates to the instrumented class. I still think it's useful to have this here because there are many other implementations of PSR-17 that we may want to support in the future.

As for the log level, I can drop it, yes.